### ISSUE TITLE
Restart loop for application compiling assets at startup.

### ISSUE THREAD

**Issue by eroldan** [vapor-ware/ksync#180]
Why ksync restarts containers when detects a file change inside the container? In my case, this creates a restart loop when the app starts and generates new files. I'm doing something wrong?

Here is my test to verify that a change inside container triggers a restart:
```
ksync create --name eroldan --selector=app=skc-api-smartclause --local-read-only --reload=true $(pwd)/edu /edu/
sleep 60
kubectl exec -it skc-api-smartclause-8f476b59b-4j9zh touch /edu/blah
```

Then I got:
```
Jun 02 00:14:33 ip-172-31-22-182 sudo[7662]: time="2018-06-02T00:14:33Z" level=info msg=updating pod=skc-api-smartclause-8f476b59b-4j9zh spec=eroldan
Jun 02 00:14:33 ip-172-31-22-182 sudo[7662]: time="2018-06-02T00:14:33Z" level=info msg="update complete" pod=skc-api-smartclause-8f476b59b-4j9zh spec=eroldan
Jun 02 00:14:33 ip-172-31-22-182 sudo[7662]: time="2018-06-02T00:14:33Z" level=info msg="issuing reload" pod=skc-api-smartclause-8f476b59b-4j9zh spec=eroldan
Jun 02 00:14:33 ip-172-31-22-182 sudo[7662]: time="2018-06-02T00:14:33Z" level=info msg=reloaded pod=skc-api-smartclause-8f476b59b-4j9zh spec=eroldan
```

**Comment by timfallmk** [comment]
@eroldan It _should_ roll the container when new files are changed, are you saying manually editing files on the container side starts a reload loop? If so is it only on the remote side?

**Comment by eroldan** [comment]
The scope of this issue is only for remote side. Specifically, if some files are added to the watched directory (manually or by the application) the container is rolled. This cause infinite restarts loops when the application compiles assets at startup or create a tmp file, etc.
For example
```
kubectl run kstest --image busybox -- sh -c 'mkdir -p /srv; cd /srv; while true; do echo "$(date)" | tee $(date +%T); sleep 30; done'
mkdir srv; ksync create --selector=run=kstest $(pwd)/srv /srv
```
The container is restarted endlessly.
I would like to disable this "roll when container fs changes" behaviour and question if this is a good default behaviour.

Versions:
```
ubuntu@ip-172-31-22-182:~$ ksync version
ksync:
        Version:    Release
        Go Version: go1.10.2
        Git Commit: 460cc64
        Git Tag:    0.1.5-hotfix
        Built:      Tue May 22 22:28:18 UTC 2018
        OS/Arch:    linux/amd64
service:
        Version:    Release
        Go Version: go1.10.2
        Git Commit: 460cc64
        Git Tag:    0.1.5-hotfix
        Built:      Wed May 23 00:26:27 UTC 2018
```

Thank you!

**Comment by timfallmk** [comment]
Does this happen if you disable restarts when creating the spec?

**Comment by eroldan** [comment]
I've added the `--restart=Never` flag to the run command and it becomes:
```kubectl run kstest --image busybox --restart=Never -- sh -c 'mkdir -p /srv; cd /srv; while true; do echo "$(date)" | tee $(date +%T); sleep 30; done'```.

I don't see any change on the ksync behaviour. Container keeps rolling. (I'm checking it with `kubectl logs ksync-b8g65 -n kube-system -c ksync -f`).

**Comment by timfallmk** [comment]
I believe restart is a bool. I’ll check all of this in the morning

**Comment by timfallmk** [comment]
@eroldan I believe the correct command would be:
```shell
ksync create --reload=false blah blah
```

**Comment by eroldan** [comment]
The flag "--reload=false" will completely disable restarts when changes are detected on local or remote ends and that isn't what I want.
I want automatic restarts when **local** changes are detected, but **not when remote** changes occurr (by manual edits on container or application generated files).

In other words:
some file changes on my laptop -> container rolls
some file changes on the container -> nothing happens

**Comment by timfallmk** [comment]
@eroldan So let me see if I'm understanding you here. What you want is:

1. Remote container filesystem reflects changes on the local filesystem, both adds and deletes
2. Local filesystem reflects changes on the remote container filesystem, both adds and deletes
3. Local changes cause the container to reload (required for how we do file updates)
4. Remote changes do not cause container to reload

Is that correct?

**Comment by eroldan** [comment]
@timfallmk Not completely correct. Your listed point 2 is not a requirement for me, altough it is OK if that happens.

**Comment by eroldan** [comment]
@grampelberg It is happening for me. Can you test it too? I've provided my test case some messages above. https://github.com/vapor-ware/ksync/issues/180#issuecomment-394939614

**Comment by grampelberg** [comment]
Nope, I was wrong. https://github.com/vapor-ware/ksync/blob/master/pkg/ksync/folder.go#L249

Looks like we'll need a new flag maybe? I'm not sure what the correct default behavior should be. I guess since we have `--local-read-only` we should also have `--local-restart-only` or something along those lines. @timfallmk what do you think?

**Comment by timfallmk** [comment]
Wouldn't it make sense for the remote to only roll on changes _pushed_ from the local? I'm not sure which is harder though, making the restart sequence aware of where a change came from or adding a new flag. Thoughts: @grampelberg

**Comment by timfallmk** [comment]
I agree with you generally, although I'm not sure why you would want updating files in the container manually to roll the container itself. In that circumstance you don't need to restart the container to have it pick up the changes, so I assume this would be for doing some sort of development directly in the container?

Would the `reload` command we have now suffice for this purpose? https://github.com/vapor-ware/ksync/pull/185

**Comment by grampelberg** [comment]
No, #185 reloads syncthing and not the actual container remotely.

**Comment by kameshsampath** [comment]
@grampelberg  - any update on this issue. I am facing the restart loop

**Comment by kameshsampath** [comment]
for now i overcame the issue with `reload=false`

**Comment by alok87** [comment]
It works when `reload=false`

A really nice experience would be when reload works flawlessly and very smoothly for the user.
Auto reload is a very nice feature to have.

I was testing the reload functionality with a node application:

I have a node application in which:
- i made a single line code change
```
$ git diff bin/server.js
-    status: "OK"
+    status: "KSYNC-TEST"
```
- I copied my change and built the code to generate the required files which will get synced to my container
`cp /tmp/server.js bin/server.js && npm run build:server`

- When i run the above command for about 30-40 seconds continuous update, updated and reloaded is seen in the `ksync watch`
`updating`
`updated`
`reloading`
`reloaded`

I am not really sure at present why exactly so many updates are happening for a single line changed file and `dist/` folder is also new and it is being copied.

I think because of many copies and sync happening not in one go - multiple reloads are happening. There should be some configurable buffer time flag before which reload should happen (some time given for sync to happen), this can help make the reload experience a lot better which is very weird and the service disruption happens because of it.

**Comment by grampelberg** [comment]
Here's the code that does it:

https://github.com/vapor-ware/ksync/blob/master/pkg/ksync/folder.go#L190

There's a time limit on there and it only reloads for local -> remote files, not remote -> local. It'd be easy to make the limit configurable, add a filter for what files constitute a reload or make sync uni-directional.

You can also up the log level for debug, which is pretty verbose, and watch the syncthing logs. My suspicion is that something locally is fighting with files getting synced back from the remote.
